modules/transformation.py

```diff
@@ -0,0 +1,500 @@
+            if cc and ('must-revalidate' in cc or 'no-cache' in cc or 'no-store' in cc):
+                # These directives dictate the response can NOT be cached.
+                exp_age = 0
+            elif cc and re.match(r"max-age=\d+", cc):
+                try:
+                    exp_age = int(re.findall(r"\d+", cc)[0])
+                except Exception:
+                    # TODO compare results from old and new pipeline for these errors
+                    logging.warning(f"Unable to parse max-age, cc:{cc}", exc_info=True)
+            elif 'resp_expires' in response_headers:
+                # According to RFC 2616 ( http://www.w3.org/Protocols/rfc2616/rfc2616-sec13.html#sec13.2.4 ):
+                #     freshness_lifetime = expires_value - date_value
+                # If the Date: response header is present, we use that.
+                # Otherwise, fall back to $startedDateTime which is based on the client so might suffer from clock skew.
+                try:
+                    start_date = date_parser.parse(response_headers.get('resp_date')[0]).timestamp() \
+                        if 'resp_date' in response_headers \
+                        else ret_request['startedDateTime']
+                    end_date = date_parser.parse(response_headers['resp_expires'][0]).timestamp()
+                    # TODO try regex to resolve issues parsing apache ExpiresByType Directive
+                    #   https://httpd.apache.org/docs/2.4/mod/mod_expires.html#expiresbytype
+                    # end_date = date_parser.parse(
+                    #   re.findall(r"\d+", response_headers['resp_expires'][0])[0]).timestamp()
+                    exp_age = end_date - start_date
+                except Exception:
+                    logging.warning(f"Could not parse dates. "
+                                    f"start=(resp_date:{response_headers.get('resp_date')},"
+                                    f"startedDateTime:{ret_request.get('startedDateTime')}), "
+                                    f"end=(resp_expires:{response_headers.get('resp_expires')}), "
+                                    f"status_info:{status_info}")
+
+            ret_request.update({
+                'expAge': int(max(exp_age, 0))
+            })
+
+            # NOW add all the headers from both the request and response.
+            ret_request.update({k: ", ".join(v) for k, v in request_headers.items()})
+
+            # TODO consider doing this sooner? why process if status is bad?
+            # wrap it up
+            first_req = False
+            first_html = False
+            if not first_url:
+                if (400 <= status <= 599) or 12000 <= status:
+                    logging.error(
+                        "ERROR($gPagesTable pageid: {}): The first request ({}}) failed with status {}}.",
+                        pageid, url, status
+                    )
+                    return None
```

Comment:
That appears to be correct, here is the original PHP implementation
https://github.com/HTTPArchive/legacy.httparchive.org/blob/de08e0c7c94a7da529826f0a4429a9d28b8fdf5e/bulktest/batch_lib.inc#L670

Reply:
Correct, the legacy pipeline threw away tests that were unsuccessful.  The "size" of 404's may be deceptive because WPT stops the test as soon as it sees the navigation page failed and doesn't wait for a full 404 page to load.  Both can be changed but then the analysis gets a bit more complicated

Reply:
Cool. Resolving.